etcdserver/api/v3rpc/maintenance.go

```diff
@@ -45,19 +46,45 @@ type RaftStatusGetter interface {
 	Leader() types.ID
 }
 
+type AuthGetter interface {
+	AuthInfoFromCtx(ctx context.Context) (*auth.AuthInfo, error)
+	AuthStore() auth.AuthStore
+}
+
 type maintenanceServer struct {
 	rg  RaftStatusGetter
 	kg  KVGetter
 	bg  BackendGetter
 	a   Alarmer
+	ag  AuthGetter
 	hdr header
 }
 
 func NewMaintenanceServer(s *etcdserver.EtcdServer) pb.MaintenanceServer {
-	return &maintenanceServer{rg: s, kg: s, bg: s, a: s, hdr: newHeader(s)}
+	return &maintenanceServer{rg: s, kg: s, bg: s, a: s, ag: s, hdr: newHeader(s)}
+}
+
+func (ms *maintenanceServer) isAuthenticated(ctx context.Context) error {
+	authInfo, err := ms.ag.AuthInfoFromCtx(ctx)
+	if err != nil {
+		return err
+	}
+
+	as := ms.ag.AuthStore()
+	err = as.IsAdminPermitted(authInfo)
```

Comment:
I'll change it after finishing another ongoing PR: https://github.com/coreos/etcd/pull/6903